ee/clickhouse/queries/clickhouse_stickiness.py

```diff
@@ -1,10 +1,133 @@
+from posthog.queries.base import BaseQuery, determine_compared_filter
+from posthog.utils import relative_date_parse
+
+STICKINESS_SQL = """
+    SELECT countDistinct(person_id), day_count FROM (
+         SELECT person_distinct_id.person_id, countDistinct(toDate(timestamp)) as day_count
+         FROM events
+         LEFT JOIN person_distinct_id ON person_distinct_id.distinct_id = events.distinct_id
+         WHERE team_id = {team_id} AND event = '{event}' {filters} {parsed_date_from} {parsed_date_to}
+         GROUP BY person_distinct_id.person_id
+    ) GROUP BY day_count ORDER BY day_count
+"""
+
+STICKINESS_ACTIONS_SQL = """
+    SELECT countDistinct(person_id), day_count FROM (
+         SELECT person_distinct_id.person_id, countDistinct(toDate(timestamp)) as day_count
+         FROM events
+         LEFT JOIN person_distinct_id ON person_distinct_id.distinct_id = events.distinct_id
+         WHERE team_id = {team_id} AND id IN ({actions_query}) {filters} {parsed_date_from} {parsed_date_to}
+         GROUP BY person_distinct_id.person_id
+    ) GROUP BY day_count ORDER BY day_count
+"""
 
 
 class ClickhouseStickiness(BaseQuery):
+    def _serialize_entity(
+        self, entity: Entity, filter: Filter, team: Team, label_note: str = ""
+    ) -> List[Dict[str, Any]]:
+        serialized: Dict[str, Any] = {
+            "action": entity.to_dict(),
+            "label": "{}{}".format("{} — ".format(label_note), entity.name),
+            "count": 0,
+            "data": [],
+            "labels": [],
+            "days": [],
+        }
+
+        result = self._format_stickiness_query(entity, filter, team)
+
+        new_dict = copy.deepcopy(serialized)
+        new_dict.update(result)
+        return [new_dict]
+
+    def _format_stickiness_query(self, entity: Entity, filter: Filter, team: Team) -> List[Dict[str, Any]]:
+        if not filter.date_to or not filter.date_from:
+            raise ValueError("_stickiness needs date_to and date_from set")
+        range_days = (filter.date_to - filter.date_from).days + 2
+
+        parsed_date_from, parsed_date_to = parse_timestamps(filter=filter)
+        prop_filters, prop_filter_params = parse_prop_clauses("id", filter.properties, team)
+
+        params: Dict = {"team_id": team.pk}
+        params = {**params, **prop_filter_params}
+        if entity.type == TREND_FILTER_TYPE_ACTIONS:
+            action = Action.objects.get(pk=entity.id)
+            action_query, action_params = format_action_filter(action)
+            if action_query == "":
+                return []
+
+            params = {**params, **action_params}
+            content_sql = STICKINESS_ACTIONS_SQL.format(
+                team_id=team.pk,
+                actions_query=action_query,
+                parsed_date_from=(parsed_date_from or ""),
+                parsed_date_to=(parsed_date_to or ""),
+                filters="AND id IN {filters}".format(filters=prop_filters) if filter.properties else "",
+            )
```

Comment:
This seems susceptible to SQL injection attacks (particularly through filters). Do we have protections around that?

Reply:
It's definitely susceptible for this! I left it as is though because all other/existing clickhouse queries currently are equally vulnerable right now. This is definitely something to fix before any of this goes live.

Reply:
yeah we will be doing a full parameterization before deploying. the syntax is just `%(variable_name)s`. i've done it in some spots but also just directly formatted in others out of convenience while i was experimenting